crates/ripemd160/src/util.rs

```diff
@@ -0,0 +1,211 @@
+use bellpepper::gadgets::uint32::UInt32;
+use bellpepper_core::{boolean::Boolean, ConstraintSystem, SynthesisError};
+use ff::PrimeField;
+
+pub fn swap_byte_endianness(bits: &[Boolean]) -> Vec<Boolean> {
+    assert!(bits.len() % 8 == 0);
+    let mut modified_bits = vec![];
+    for i in 0..bits.len() / 8 {
+        for j in 0..8 {
+            modified_bits.push(bits[i * 8 + 7 - j].clone());
+        }
+    }
+    modified_bits
+}
+
+pub fn uint32_rotl(a: &UInt32, by: usize) -> UInt32 {
+    assert!(by < 32usize);
+    a.rotr(32 - by)
+}
+
+fn triop<Scalar, CS, U>(
+    mut cs: CS,
+    a: &UInt32,
+    b: &UInt32,
+    c: &UInt32,
+    circuit_fn: U,
+) -> Result<UInt32, SynthesisError>
```

Comment:
The use of `into_bits` will force a clone, so you're better off asking for a, b, c by value here and let the caller manage the cloning.
See also [C-CALLER-CONTROL](https://rust-lang.github.io/api-guidelines/flexibility.html#c-caller-control)

Reply:
Thanks for the suggestion. I have made this change.